xdsl/transforms/canonicalization_patterns/arith.py

```diff
@@ -29,16 +52,49 @@ def match_and_rewrite(
             and isinstance(op.rhs.owner, arith.Constant)
             and isa(l := op.lhs.owner.value, builtin.AnyFloatAttr)
             and isa(r := op.rhs.owner.value, builtin.AnyFloatAttr)
+            and (cnst := _fold_const_operation(type(op), l, r))
+        ):
+            rewriter.replace_matched_op(cnst)
+
+
+class FoldConstsByReassociation(RewritePattern):
+    """
+    Rewrites a chain of
+        `(const1 <op> var) <op> const2`
+    as
+        `folded_const <op> val`
+
+    The op must be associative and have the `fastmath<reassoc>` flag set.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: arith.Addf | arith.Mulf, rewriter: PatternRewriter, /
+    ):
+        if isinstance(op.lhs.owner, arith.Constant):
+            const1, val = op.lhs.owner, op.rhs
+        else:
+            const1, val = op.rhs.owner, op.lhs
+
+        if (
+            not isinstance(const1, arith.Constant)
+            or len(op.result.uses) != 1
+            or not isinstance(u := list(op.result.uses)[0].operation, type(op))
+            or not isinstance(
+                const2 := u.lhs.owner if u.rhs == op.result else u.rhs.owner,
+                arith.Constant,
+            )
+            or op.fastmath is None
+            or u.fastmath is None
+            or arith.FastMathFlag.REASSOC not in op.fastmath.flags
+            or arith.FastMathFlag.REASSOC not in u.fastmath.flags
```

Comment:
It's just that my use-cases are covered by ops that happen to be commutative. In my understanding, `(var - cst) - cst` could just as easily be reassociated and simplified, although non-commutative ops have more patterns that would need to be considered.

The [docs](https://llvm.org/docs/LangRef.html#rewrite-based-flags) are a bit light on the detail, though mention that this can have a significant effect on fp results. I understand llvm has a reassociate pass, for which I can't find an equivalent in mlir.